default/python/AI/universe/system_network.py

```diff
@@ -35,6 +36,23 @@ def _systems_connected(system1: SystemId, system2: SystemId) -> bool:
     if system1 == INVALID_ID:
         return False
 
+    # optimization: We cache a set of systems connected to the capital
+    # system in a single DFS and check if the systems are both in that set.
+    # This allows us to avoid a BFS for each pair of systems.
+    # Only in case neither of the system connects to the capital, we need to
+    # run the BFS.
+    connected_system_set = systems_connected_to_system(get_capital_sys_id())
+    sys1_connected_to_capital = system1 in connected_system_set
+    sys2_connected_to_capital = system2 in connected_system_set
+    if sys1_connected_to_capital and sys2_connected_to_capital:
+        # both connected to the capital - so must be connected to each other
+        return True
+
+    if sys1_connected_to_capital != sys2_connected_to_capital:
+        # Only one connected to the capital - can't be connected to each other
+        return False
```

Comment:
Let's simplify code a bit:

```python
#both systems are in the same cluster
return system1 in connected_system_set and system2 in connected_system_set
```

Reply:
This unfortunately does not cover the third return value where we still run the BFS.
Following would work but I find it comes at a cost of readability, so prefer the current version.
```
if system1 not in connected_system_set and system2 not in connected_system_set:
    return fo.getUniverse().systemsConnected(system1, system2, fo.empireID())  # run full BFS

return system1 in connected_system_set and system2 in connected_system_set
```

Reply:
My bad.  I usually extract such code to a function. Only in that context, my suggestion makes sense.

```python
def is_in_same_cluster(sys1, sys2):
    ...
    return system1 in connected_system_set and system2 in connected_system_set

def _systems_connected...
    ...
    if is_in_same_cluster(sys1, sys2):
        return True
    ...